src/createAnimatedComponent/JSPropsUpdater.ts

```diff
@@ -0,0 +1,153 @@
+    }
+    this._reanimatedEventEmitter = new NativeEventEmitter(reanimatedModule);
+  }
+
+  public addOnJSPropsChangeListener(
+    animatedComponent: React.Component<unknown, unknown>
+  ) {
+    const viewTag = findNodeHandle(animatedComponent);
+    JSPropsUpdaterPaper._tagToComponentMapping.set(viewTag, animatedComponent);
+    if (JSPropsUpdaterPaper._tagToComponentMapping.size === 1) {
+      const listener = (data: ListenerData) => {
+        const component = JSPropsUpdaterPaper._tagToComponentMapping.get(
+          data.viewTag
+        );
+        component?._updateFromNative(data.props);
+      };
+      this._reanimatedEventEmitter.addListener(
+        'onReanimatedPropsChange',
+        listener
+      );
+    }
+  }
+
+  public removeOnJSPropsChangeListener(
+    animatedComponent: React.Component<unknown, unknown>
+  ) {
+    const viewTag = findNodeHandle(animatedComponent);
+    JSPropsUpdaterPaper._tagToComponentMapping.delete(viewTag);
+    if (JSPropsUpdaterPaper._tagToComponentMapping.size === 0) {
+      this._reanimatedEventEmitter.removeAllListeners(
+        'onReanimatedPropsChange'
+      );
+    }
+  }
+}
+
+class JSPropsUpdaterFabric implements JSPropsUpdater {
+  private static _tagToComponentMapping = new Map();
+  private static isInitialized = false;
+
+  constructor() {
+    if (nativeShouldBeMock()) {
+      return;
+    }
+    if (!JSPropsUpdaterFabric.isInitialized) {
+      const updater = (viewTag: number, props: unknown) => {
+        const component =
+          JSPropsUpdaterFabric._tagToComponentMapping.get(viewTag);
+        component?._updateFromNative(props);
+      };
+      runOnUIImmediately(() => {
+        'worklet';
+        global.updateJSProps = (viewTag: number, props: unknown) => {
+          runOnJS(updater)(viewTag, props);
+        };
+      })();
+      JSPropsUpdaterFabric.isInitialized = true;
+    }
+  }
+
+  public addOnJSPropsChangeListener(
+    animatedComponent: React.Component<unknown, unknown>
+  ) {
+    if (!JSPropsUpdaterFabric.isInitialized) {
+      return;
+    }
+    const viewTag = findNodeHandle(animatedComponent);
+    JSPropsUpdaterFabric._tagToComponentMapping.set(viewTag, animatedComponent);
+  }
+
+  public removeOnJSPropsChangeListener(
+    animatedComponent: React.Component<unknown, unknown>
+  ) {
+    if (!JSPropsUpdaterFabric.isInitialized) {
+      return;
+    }
+    const viewTag = findNodeHandle(animatedComponent);
+    JSPropsUpdaterFabric._tagToComponentMapping.delete(viewTag);
+  }
+}
+
+class JSPropsUpdaterWeb implements JSPropsUpdater {
+  public addOnJSPropsChangeListener(
+    _animatedComponent: React.Component<unknown, unknown>
+  ) {
+    // noop
+  }
+
+  public removeOnJSPropsChangeListener(
+    _animatedComponent: React.Component<unknown, unknown>
+  ) {
+    // noop
+  }
+}
+
+function getJSPropsUpdater():
+  | typeof JSPropsUpdaterWeb
+  | typeof JSPropsUpdaterFabric
+  | typeof JSPropsUpdaterPaper {
```

Comment:
Can we assign this type above?
```ts
type JSPropsUpdater =
  | typeof JSPropsUpdaterWeb
  | typeof JSPropsUpdaterFabric
  | typeof JSPropsUpdaterPaper;

function getJSPropsUpdater(): JSPropsUpdater {
  // ...
}
```
Also, do we really need to specify the return type here?

Reply:
I think we should because Typescript can't recognise correct type of JSPropsUpdater. Here is example:
Without type specification:
![Screenshot 2023-10-13 at 10 45 02](https://github.com/software-mansion/react-native-reanimated/assets/36106620/553201aa-5fc4-4a3f-bff5-a80e649209df)
With type specification:
![Screenshot 2023-10-13 at 10 44 30](https://github.com/software-mansion/react-native-reanimated/assets/36106620/f2f01670-947f-49e2-9cea-2b2799a14c68)